Access Token IDs have less than 128 bits of entropy

douglaspalmer opened the pull request:
Closes #38663

<!---
Please read https://github.com/keycloak/keycloak/blob/main/CONTRIBUTING.md and follow these guidelines when contributing to Keycloak
-->

douglaspalmer committed:
Access Token IDs have less than 128 bits of entropy

Closes #38663

Signed-off-by: Douglas Palmer <dpalmer@redhat.com>

rmartinc commented on adapters/saml/core/src/main/java/org/keycloak/adapters/saml/rotation/SamlDescriptorPublicKeyLocator.java:
```diff
@@ -179,7 +180,10 @@ private synchronized <T> Key refreshCertificateCacheAndGet(T key, Map<T, Key> ca
             } else {
                 final X500Principal principal = x509certificate.getSubjectX500Principal();
-                String name = (principal == null ? "unnamed" : principal.getName()) + "@" + x509certificate.getSerialNumber() + "$" + UUID.randomUUID();
+                SecureRandom random = new SecureRandom();
+                byte[] bytes = new byte[16];
+                random.nextBytes(bytes);
```
Would not be better to use `SecretGenerator.randomBytes(16)`? It's in the common package so I think you can use in any other package.

rmartinc commented on adapters/saml/core/src/main/java/org/keycloak/adapters/saml/rotation/SamlDescriptorPublicKeyLocator.java:
```diff
@@ -179,7 +180,10 @@ private synchronized <T> Key refreshCertificateCacheAndGet(T key, Map<T, Key> ca
+                SecureRandom random = new SecureRandom();
+                byte[] bytes = new byte[16];
+                random.nextBytes(bytes);
+                String name = (principal == null ? "unnamed" : principal.getName()) + "@" + x509certificate.getSerialNumber() + "$" + Arrays.toString(bytes);
```
Using `Arrays.toString` would create something like `[56, -120,...]` which is a horrible id IMHO. Why don't you create a new utility method in the `SecretGenerator` class to generate the random array and return it in hexa?

```java
    public static String randomBytesHexa(int length) {
        final StringBuilder sb = new StringBuilder();
        for (byte b : randomBytes(length) {
            sb.append(Character.forDigit((b >> 4) & 0xF, 16));
            sb.append(Character.forDigit((b & 0xF), 16));
        }
        return sb.toString();
    }
```

If you prefer to make it more similar to the UUID format you can also add the hyphen `-` after 4 or 8 bytes.

rmartinc reviewed (changes requested):
Thanks @douglaspalmer! WFYT about using the `SecretGenerator` if possible? And I think using `Arrays.toString` is not OK, we can use hexa for example.

rmartinc reviewed (approved):
Thganks @douglaspalmer! LGTM! Seeing the `SecretGenerator` now I would have preferred to not make the method static and use the `getInstance` but it's a minor detail. It's OK to me.

douglaspalmer commented:
> Thganks @douglaspalmer! LGTM! Seeing the `SecretGenerator` now I would have preferred to not make the method static and use the `getInstance` but it's a minor detail. It's OK to me.

Done, can you please review again?

rmartinc reviewed (approved):
Thanks @douglaspalmer! LGTM!

rmartinc commented:
@douglaspalmer Rebase this to see if CI runs OK.
@mposolda For me this is OK, now the IDs are just the hexa with no hyphens (so previously `58e0a7d7-eebc-11d8-9669-0800200c9a66` and now just the hexa `4e4d4046ab29803d1f38273d0938cc30`). But I think the ID is OK.

mposolda commented:
@douglaspalmer Few minor points:

- The issue https://github.com/keycloak/keycloak/issues/38663 mentions access-token IDs (and few other things in the description). But it seems this PR handles various secrets, but does not handle access-token IDs particularly, which are mentioned as main part in the issue? As access-token IDs (and similarly IDTokens or refreshToken Ids) are generated with `KeycloakModelUtils.generateId()` , which itself likely cannot be changed as it is used for many other things (EG. generating IDs for DB entities). See for example `TokenManager.initToken()` where is generated ID of the access-token.  There are few other methods on `TokenManager` for ID tokens and refresh tokens). It can be good to update those places as well.

- How about adding some method on `SecretGenerator` like:

```
public String generateSecureID() {
    SecretGenerator.getInstance().randomBytesHex(16);
}
```
It means that in 5 years or so, when there are new "security best practices", we will not need to change tons of places in the codebase and increase 16 to 32 for example, but we might be able to change just one place :-) WDYT?

ahus1 commented:
@douglaspalmer - IMHO changing the IDs of the tokens could justify an entry in the notable changes in `changes-26_3_0.adoc`. While it is an internal behavior, maybe some people are depending on it in some kind of way. 

If you want this PR to be merged rather sooner than later, the updated docs can be handled in a follow-up PR. Leaving that up to you,  @mposolda  and @rmartinc  to decide.

douglaspalmer commented:
> @douglaspalmer - IMHO changing the IDs of the tokens could justify an entry in the notable changes in `changes-26_3_0.adoc`. While it is an internal behavior, maybe some people are depending on it in some kind of way.
> 
> If you want this PR to be merged rather sooner than later, the updated docs can be handled in a follow-up PR. Leaving that up to you, @mposolda and @rmartinc to decide.

I changed the PR yesterday and added dashes to the IDs. So the change should be invisible now.

mposolda reviewed (approved):
@douglaspalmer Sounds good, Thanks. Considering that the entropy is bigger, however the length of the ID and also the format is quite the same, I do not think that this worth to be documented.
@ahus1 If you disagree, feel free to create follow-up issue for this and we can eventually document for 26.3.
